doc/crdt-details.md

```diff
@@ -0,0 +1,1007 @@
+# The Xi Text Engine CRDT
+
+This document contains a detailed description of Xi's text Conflict-free Replicated Data Type (CRDT) as implemented in the `xi-rope` crate. It describes the actual Rust data structures and algorithms used, because the primary novelty and difficulty of this CRDT is in the optimized representation that allows for better time and memory complexity. If you want an overview of the motivation behind using a CRDT and a conceptual description of what the CRDT does see `crdt.md`.
+
+## Table of Contents
+
+- [Motivation](#motivation): Why Xi's CRDT is the way it is.
+- [Representation](#representation): Describes the representation Xi uses to implement the CRDT in a memory and time efficient way.
+- [Operations](#operations): Describes all the operations implemented on the representation to allow it to support undo, asynchronous edits, distributed synchronization and more.
+
+## Motivation
+
+The Xi CRDT attempts to have a number of properties that many other asynchronous text editing solutions don't:
+
+- Usable without a central server: Even though the Fuchsia Ledger does sync to a cloud server, it can't actually inspect the data or perform operations, so conflict resolution must be possible on every device independently.
+- Support large documents: The memory complexity of the representation and the time complexity of common operations should be low enough to support very large text documents like books or long code files.
+- Support long histories: Similarly, documents that have gone through many edits should be efficient to edit both in time and memory.
+
+As of the time this document was written, it satisfies all of these properties to some extent, but some operations and representations are not as memory and time efficient as we'd like. However, everything has been designed with a few more key optimizations in mind. Those optimizations should bring the memory and time complexity down to where we want.
+
+### Transform Property 2 (TP2) and Operational Transforms
+
+Operational Transformation (OT) is a common way to implement asynchronous text editing. It works by sending *operations* like inserts and deletes between peers and transforming them to apply to the current text. Unfortunately many implementations of OT have a problem where they don't always preserve ordering when text is deleted. For example see the following diagram showing 3 peers sending edits between each other ending up in an inconsistent state:
+
+![TP2 Problem](img/tp2.png)
+
+Acting consistently in cases like this is called having "Transform Property 2". The difficulty of solving this with Operational Transformation can be solved by having a central server do all the transformation so that although the order over deletes is not preserved at least the result can be made consistent. This is what Google Docs and many other collaborative editing systems do.
```

Comment:
I'd cite [Operational Transformation in Real-Time Group Editors:
Issues, Algorithms, and Achievements](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.53.933&rep=rep1&type=pdf) by Sun & Ellis, 1998, here. I'm ok either with references being simple inline links, or having a references section at the end.

The discussion of TP2 vs what most systems do seems correct, but feels like it might be confusing. What's in my mind is, "there are at least two approaches to the TP2 problem. One is to use a central server that serializes all edits. With this approach, the ordering relative to deleted text may not be preserved, but at least all clients will converge to the same state."